### PR TITLE
Fix ALPN SNI Proxy TLS termination for DB connections

### PR DESCRIPTION
**Pull request by smallinsky**
# What 

- Fix psql client connection - psql takes only DNS name into consideration during host name validation thus connecting to "127.0.0.1" will result with following psql error: 
```
psql: error: server certificate for "ice-berg.dev" (and 6 other names) does not match host name "127.0.0.1"
```
- Fix Mongo client connection - Terminate DB TLS client connection using cluster cert valid for `localhost` and `127.0.0.1` instead of web  proxy cert where local domain can be missing from SANs.

### CODE REVIEW

File: tool/tsh/db.go [hunk 1]
```diff
@@ -335,7 +335,13 @@ func onDatabaseConnect(cf *CLIConf) error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		opts = append(opts, WithLocalProxy(addr.Host(), addr.Port(0)))
+		host := addr.Host()
+		if database.Protocol == defaults.ProtocolPostgres {
+			// In case of psql connection the cli takes only domain names into consideration thus
+			// connecting to 127.0.0.1 will fail with "does not match host name "127.0.0.1" psql error.
+			host = "localhost"
+		}
```
Comment (r0mant) [inline tool/tsh/db.go:343]: Instead of treating Postgres as a special case, can we just always connect to local proxy over `localhost` instead?

Comment (smallinsky) [reply tool/tsh/db.go]: This work for mongo and postgres but for mysql localhost is a special value and when mysql client get localhost is try to by default to use Unix domain socket for connection:
So for example: 
`mysql --user root --host localhost -p 3306`
will return a error:
```
Can't connect to local MySQL server through socket '/tmp/mysql.sock'
```
but when `127.0.0.1` address is used it works: 
`mysql --user root --host 127.0.0.1 -p 3306`

But we can bypass this by providing the "--protocol=TCP" flag

File: integration/proxy_test.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib"
 	"github.com/gravitational/teleport/lib/service"
-	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	common2 "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
```
Comment (r0mant) [inline integration/proxy_test.go:36]: Let's do `alpncommon` or something like that rather than `2` suffix.

File: lib/auth/auth.go [hunk 3]
```diff
@@ -3345,3 +3349,60 @@ func isHTTPS(u string) error {
 
 	return nil
 }
+
+// AddClusterCAPoolsToTLSConfig return a copy of tls.Config and set the  ClientCAs
+// to the cluster CA Pool.
```
Comment (r0mant) [inline lib/auth/auth.go:3354]:
```suggestion
// AddClusterCAPoolsToTLSConfig returns a TLS hello callback that returns a copy of the provided
// TLS config with client CAs pool of the specified cluster.
```

File: lib/auth/auth.go [hunk 4]
```diff
@@ -3345,3 +3349,60 @@ func isHTTPS(u string) error {
...
+func AddClusterCAPoolsToTLSConfig(tlsConfig *tls.Config, ap AccessPoint, currentClusterName string, log logrus.FieldLogger) func(*tls.ClientHelloInfo) (*tls.Config, error) {
```
Comment (r0mant) [inline lib/auth/auth.go:3355]: Nit: This does not really add anything by itself but returns a function that does it rather. Maybe a name like `WithClusterCAs` or something like that would work a little better?

File: lib/auth/auth.go [hunk 5]
```diff
@@ -3345,3 +3349,60 @@ func isHTTPS(u string) error {
...
+			if err != nil {
+				if !trace.IsNotFound(err) {
+					log.Debugf("Ignoring unsupported cluster name name %q.", info.ServerName)
+					clusterName = ""
+				}
+			}
+		}
+		pool, err := ClientCertPool(ap, clusterName)
+		if err != nil {
+			log.Errorf("failed to retrieve client pool: %v", trace.DebugReport(err))
```
Comment (r0mant) [inline lib/auth/auth.go:3371]:
```suggestion
			log.WithError(err).Errorf("Failed to retrieve client pool for %q.", clusterName)
```

File: lib/auth/auth.go [hunk 6]
```diff
@@ -3345,3 +3349,60 @@ func isHTTPS(u string) error {
...
+		// the current cluster CA. In the unlikely case where it's wrong, the
+		// client will be rejected.
+		var totalSubjectsLen int64
+		for _, s := range pool.Subjects() {
+			// Each subject in the list gets a separate 2-byte length prefix.
+			totalSubjectsLen += 2
+			totalSubjectsLen += int64(len(s))
+		}
+		if totalSubjectsLen >= int64(math.MaxUint16) {
+			log.Debugf("number of CAs in client cert pool is too large (%d) and cannot be encoded in a TLS handshake; this is due to a large number of trusted clusters; will use only the CA of the current cluster to validate", len(pool.Subjects()))
```
Comment (r0mant) [inline lib/auth/auth.go:3395]:
```suggestion
			log.Debugf("Number of CAs in client cert pool is too large (%d) and cannot be encoded in a TLS handshake; this is due to a large number of trusted clusters; will use only the CA of the current cluster to validate.", len(pool.Subjects()))
```

File: lib/auth/auth.go [hunk 7]
```diff
@@ -3345,3 +3349,60 @@ func isHTTPS(u string) error {
...
+
+			pool, err = ClientCertPool(ap, currentClusterName)
+			if err != nil {
+				log.Errorf("failed to retrieve client pool: %v", trace.DebugReport(err))
```
Comment (r0mant) [inline lib/auth/auth.go:3399]:
```suggestion
				log.WithError(err).Errorf("Failed to retrieve client pool for %q.", currentClusterName)
```

File: lib/bpf/bytecode/README.md [hunk 8]
```diff
@@ -1,3 +0,0 @@
-## BPF Bytecode
```
Comment (r0mant) [inline lib/bpf/bytecode/README.md:1]: Did you remove this file on purpose?

File: lib/restrictedsession/bytecode/README.md [hunk 9]
```diff
@@ -1,3 +0,0 @@
-## BPF Bytecode
```
Comment (r0mant) [inline lib/restrictedsession/bytecode/README.md:1]: This too - did you intend to delete this?

File: lib/srv/alpnproxy/common/protocols.go [hunk 10]
```diff
@@ -50,7 +50,7 @@ const (
 	ProtocolAuth Protocol = "teleport-auth@"
 )
 
-var supportedProtocols = []Protocol{
+var SupportedProtocols = []Protocol{
```
Comment (r0mant) [inline lib/srv/alpnproxy/common/protocols.go:53]: Nit: Add godoc now that this is public.

File: lib/srv/alpnproxy/common/protocols.go [hunk 11]
```diff
@@ -63,7 +63,7 @@ var supportedProtocols = []Protocol{
 	ProtocolAuth,
 }
 
-func convProtocolsToString(protocols []Protocol) []string {
+func ConvProtocolsToString(protocols []Protocol) []string {
```
Comment (r0mant) [inline lib/srv/alpnproxy/common/protocols.go:66]: Nit: Add godoc and also, maybe "Conv" is redundant in the name? Just `ProtocolsToString` works well too.

File: lib/srv/alpnproxy/proxy.go [hunk 12]
```diff
@@ -58,6 +60,12 @@ type ProxyConfig struct {
 	// of the connection). It is set to defaults.HandshakeReadDeadline if
 	// unspecified.
 	ReadDeadline time.Duration
+	// IdentityTLSConfig is TLS cluster configuration.
+	IdentityTLSConfig *tls.Config
```
Comment (r0mant) [inline lib/srv/alpnproxy/proxy.go:64]: It's not super clear what "identity TLS config" is, could you rename the field and/or expand the comment?

File: lib/srv/alpnproxy/proxy.go [hunk 13]
```diff
@@ -212,10 +230,14 @@ func New(cfg ProxyConfig) (*Proxy, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
+	cfg.IdentityTLSConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	fn := auth.AddClusterCAPoolsToTLSConfig(cfg.IdentityTLSConfig, cfg.AccessPoint, cfg.ClusterName, cfg.Log)
+	cfg.IdentityTLSConfig.GetConfigForClient = fn
```
Comment (r0mant) [inline lib/srv/alpnproxy/proxy.go:235]: Can this be done as a part of `CheckAndSetDefaults`? It already clones the config.

File: tool/tsh/db.go [hunk 14]
```diff
@@ -335,7 +335,11 @@ func onDatabaseConnect(cf *CLIConf) error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		opts = append(opts, WithLocalProxy(addr.Host(), addr.Port(0)))
+
+		// In case of psql connection the cli takes only domain names into consideration thus
+		// connecting to 127.0.0.1 will fail with "does not match host name "127.0.0.1" psql error.
```
Comment (r0mant) [inline tool/tsh/db.go:340]:
```suggestion
		// When connecting over TLS, psql only validates hostname against presented certificate's
		// DNS names. As such, connecting to 127.0.0.1 will fail validation, so connect to localhost.
```

**Review by r0mant — APPROVED** [review]
Mostly stylistic nits.

**Comment by russjones** [comment]
@Joerger Can you review?

File: lib/restrictedsession/bytecode/README.md (hunk 9 above)
Comment (smallinsky) [reply lib/restrictedsession/bytecode/README.md:1]: No, this was not intended. Thanks.

**Review by Joerger — APPROVED** [review]

**Pull request merged**